File: 01_SOLID/03_LiskovSubstitutionPrinciple (LSP)/Main.cpp
#include <iostream>
#include <string>

// Notification base class
class Notification {
public:
    virtual void sendNotification() {
        std::cout << "Notification sent" << std::endl;
    }
    virtual ~Notification() = default;
};

// Subclass of Notification class for Email Notification
class EmailNotification : public Notification {
public:
    void sendNotification() override {
        std::cout << "Email Notification sent" << std::endl;
    }
};

// Subclass of Notification class for Text Notification
class TextNotification : public Notification {
public:
    void sendNotification() override {
        std::cout << "Text Notification sent" << std::endl;
    }
};

int main() {
    /* Replaced the Notification class object
    with one of its subclass' objects */
    Notification* notification = new EmailNotification();
    
    // Working code on the notification object
    notification->sendNotification();
    
    delete notification;
    return 0;
}

File: 01_SOLID/03_LiskovSubstitutionPrinciple (LSP)/Main.java
// Notification class
class Notification {
    // method implementing send notification functionality 
    public void sendNotification() {
        System.out.println("Notification sent");
    }
}

// Subclass of Notification class for Email Notification
class EmailNotification extends Notification {
    @Override
    public void sendNotification() {
        System.out.println("Email Notification sent");
    }
}

// Subclass of Notification class for Text Notification
class TextNotification extends Notification {
    @Override
    public void sendNotification() {
        System.out.println("Text Notification sent");
    }
}


// Main class
class Main {
    //  main method
    public static void main(String args[]) {
        /* Replaced the Notification class object
        with one of its subclass' objects */
        Notification notification = new EmailNotification();
        
        // Working code on the notification object
        notification.sendNotification();
    }
}
